core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -801,30 +799,32 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
   }
 
   def removePartitionFromReassignedPartitions(topicAndPartition: TopicAndPartition) {
-    if(controllerContext.partitionsBeingReassigned.get(topicAndPartition).isDefined) {
-      // stop watching the ISR changes for this partition
-      val partitionReassignmentIsrChangeHandler =
-        controllerContext.partitionsBeingReassigned(topicAndPartition).partitionReassignmentIsrChangeHandler
-      zkClient.unregisterZNodeChangeHandler(partitionReassignmentIsrChangeHandler.path)
+    val partitionsBeingReassigned = controllerContext.partitionsBeingReassigned.get(topicAndPartition) match {
+      case Some(reassignContext) =>
+        // stop watching the ISR changes for this partition
+        zkClient.unregisterZNodeChangeHandler(reassignContext.reassignIsrChangeHandler.path)
+        controllerContext.partitionsBeingReassigned
+      case None =>
+        // read reassigned partitions from zookeeper if this partition is not cached
+        zkClient.getPartitionReassignment.mapValues(replicas => ReassignedPartitionsContext(replicas))
```

Comment:
@junrao not sure if we need this, but looking at all the callers, the event `process` method seems like the only place where we could call this method with a partition that doesn't exist in the cache. And we would only do it if the topic is queued for deletion.

So, if someone managed to get the reassignment event to be processed after multiple topics had been queued for deletion, then we could potentially remove them all in the first `removePartitionFromReassignedPartitions` call for one of the relevant partitions. Seems harmless, but a bit confusing. Do you think this fallback is worth it?

The event code:

```scala
       val partitionsToBeReassigned = partitionReassignment.filterNot(p => controllerContext.partitionsBeingReassigned.contains(p._1))
        partitionsToBeReassigned.foreach { partitionToBeReassigned =>
          if (topicDeletionManager.isTopicQueuedUpForDeletion(partitionToBeReassigned._1.topic)) {
            error("Skipping reassignment of partition %s for topic %s since it is currently being deleted"
              .format(partitionToBeReassigned._1, partitionToBeReassigned._1.topic))
            removePartitionFromReassignedPartitions(partitionToBeReassigned._1)
          } else {
            val context = ReassignedPartitionsContext(partitionToBeReassigned._2)
            initiateReassignReplicasForTopicPartition(partitionToBeReassigned._1, context)
          }
        }